src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs

```diff
@@ -435,14 +436,98 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
-                return new X509Certificate2(Path.Combine(env.ContentRootPath, certInfo.Path), certInfo.Password);
+                var environment = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
+                if (certInfo.KeyPath != null)
+                {
+                    X509Certificate2 certificate = null;
+                    if (TryReadPemRSAKey(certInfo, environment, out var rsaKey))
+                    {
+                        var publicCertificate = new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path));
+                        certificate =  publicCertificate.CopyWithPrivateKey(rsaKey);
+                    }
+
+                    if (TryReadPemDSAKey(certInfo, environment, out var dsaKey))
+                    {
+                        var publicCertificate = new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path));
+                        certificate = publicCertificate.CopyWithPrivateKey(dsaKey);
+                    }
+
+                    if (certificate != null)
+                    {
+                        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                        {
+                            return PersistKey(certificate);
+                        }
+
+                        return certificate;
+                    }
+
+                    throw new InvalidOperationException(CoreStrings.InvalidPemKey);
+                }
+
+                return new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path), certInfo.Password);
             }
             else if (certInfo.IsStoreCert)
             {
                 return LoadFromStoreCert(certInfo);
             }
             return null;
+
+            static X509Certificate2 PersistKey(X509Certificate2 fullCertificate)
+            {
+                // We need to force the key to be persisted.
+                // See https://github.com/dotnet/runtime/issues/23749
+                var certificateBytes = fullCertificate.Export(X509ContentType.Pkcs12, "");
+                return new X509Certificate2(certificateBytes, "", X509KeyStorageFlags.DefaultKeySet);
+            }
+        }
+
+        private static bool TryReadPemRSAKey(CertificateConfig certInfo, IHostEnvironment environment, out RSA key)
+        {
+            try
+            {
+                var keyPath = Path.Combine(environment.ContentRootPath, certInfo.KeyPath);
+                var rsaKey = RSA.Create();
+                if (certInfo.Password == null)
+                {
+                    rsaKey.ImportFromPem(File.ReadAllText(keyPath));
+                }
+                else
+                {
+                    rsaKey.ImportFromEncryptedPem(File.ReadAllText(keyPath), certInfo.Password);
+                }
+                key = rsaKey;
+                return true;
+            }
+            catch
```

Comment:
try/catch is a little yuck. Is this a common pattern when loading keys?

Reply:
In the crypto layer we don't guess in new API, you have to be right.  You could use the PemEncoding.TryFind API to read the PEM header and call the correct API for the file type you noticed.

Reply:
> try/catch is a little yuck. Is this a common pattern when loading keys?

I figured out I can probably just look at the public key on the cert to make the right call.